### PR TITLE
zkvm: Implement Encodable trait

### LINKED ISSUE #305: zkvm: internal Encodable trait
Organize `serialized_len`/`encode` methods under `Encodable` trait, so we can have a blanket implementation of `encode_to_vec()` to use in a few places instead of a noisy three-liner.

### PR DESCRIPTION
**Pull request by cyberbono3**
Resolves #305 
Implemented Encodable trait for encode() and serialized_length()

### CODE REVIEW

File: zkvm/src/program.rs [hunk 1]
```diff
@@ -47,6 +47,26 @@ macro_rules! def_op {
     );
 }
 
+pub(crate) trait Encodable {
+   // Returns the serialized length of a datatype who implement this trait.
+   fn encode(&self, buf: &mut Vec<u8>);
+   /// Encodes a datatype who im0lements this trait into a buffer.
+   fn serialized_length(&self) -> usize;
```
Comment (oleganza) [inline zkvm/src/program.rs]: let's add another implemented method to factor out the common encode-to-vec pattern:

```rust
fn encode_to_vec(&self) -> Vec<u8> {
    let mut buf = Vec::with_capacity(self.serialized_length());
    self.encode(&mut buf);
    buf
}
```

Then, every time we do `buf=Vec::with_capacity(serialized_length) ... encode(&mut buf)` we can call that method instead.

File: zkvm/src/program.rs [hunk 2]
```diff
@@ -47,6 +47,30 @@ macro_rules! def_op {
     );
 }
 
+pub(crate) trait Encodable {
+    // Returns the serialized length of a datatype who implement this trait.
```
Comment (oleganza) [inline zkvm/src/program.rs]:
```suggestion
    // Encodes receiver into bytes appending them to a provided buffer.
```

File: zkvm/src/program.rs [hunk 3]
```diff
@@ -47,6 +47,30 @@ macro_rules! def_op {
...
+    fn encode(&self, buf: &mut Vec<u8>);
+    /// Encodes a datatype who im0lements this trait into a buffer.
```
Comment (oleganza) [inline zkvm/src/program.rs]:
```suggestion
    /// Returns precise length in bytes for the serialized representation of the receiver.
```

File: zkvm/src/program.rs [hunk 4]
```diff
@@ -47,6 +47,30 @@ macro_rules! def_op {
...
+    fn serialized_length(&self) -> usize;
+    /// Encodes a datatype who im0lements this trait into a bytecode array
```
Comment (oleganza) [inline zkvm/src/program.rs]:
```suggestion
    /// Encodes the receiver into a newly allocated vector of bytes.
```

File: zkvm/src/contract.rs [hunk 5]
```diff
@@ -63,8 +83,7 @@ impl Contract {
             payload,
             anchor,
         };
-        let mut buf = Vec::with_capacity(contract.serialized_length());
-        contract.encode(&mut buf);
+        let mut buf = contract.encode_to_vec();
```
Comment (oleganza) [inline zkvm/src/contract.rs]:
```suggestion
        let buf = contract.encode_to_vec();
```

buffer is not mutable in this scope

File: zkvm/src/program.rs [hunk 6]
```diff
@@ -47,6 +47,30 @@ macro_rules! def_op {
     );
 }
 
+pub(crate) trait Encodable {
```
Comment (oleganza) [inline zkvm/src/program.rs]: let's move this into `encoding.rs`?

Comment (cyberbono3) [reply zkvm/src/program.rs]: done

**Pull request merged**